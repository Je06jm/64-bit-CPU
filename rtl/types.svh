`ifndef CPU_TYPES
`define CPU_TYPES

package types;

typedef logic signed[7:0] byte_t;
typedef logic signed[15:0] short_t;
typedef logic signed[31:0] int_t;
typedef logic signed[63:0] long_t;

typedef logic unsigned[7:0] ubyte_t;
typedef logic unsigned[15:0] ushort_t;
typedef logic unsigned[31:0] uint_t;
typedef logic unsigned[63:0] ulong_t;

endpackage

`endif
