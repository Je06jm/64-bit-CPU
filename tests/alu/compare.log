VCD info: dumpfile tests/alu/compare.vcd opened for output.
VCD warning: $dumpvars: Package ($unit) is not dumpable with VCD.
VCD warning: $dumpvars: Package (instructions) is not dumpable with VCD.
VCD warning: $dumpvars: Package (types) is not dumpable with VCD.
Passed (tests/alu/compare.sv, 29 [1]): 3 does not equals 1
Passed (tests/alu/compare.sv, 30 [1]): 3 is not less then 1
Passed (tests/alu/compare.sv, 35 [2]): 2 equals 2
Passed (tests/alu/compare.sv, 36 [2]): 2 is not less then 2
Passed (tests/alu/compare.sv, 41 [3]): 4 does not equals 5
Passed (tests/alu/compare.sv, 42 [3]): 4 is less then 5
tests/alu/compare.sv:44: $finish called at 3 (1s)
