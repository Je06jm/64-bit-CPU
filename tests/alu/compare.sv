`include "tests/testing.svh"

`include "rtl/types.svh"

import types::*;

module ICompareTest();

    ulong_t a, b;
    logic zero, negitive;

    ICompare cmp(
        a, b,
        zero, negitive
    );

    initial begin
        a = 0;
        b = 0;
    end

    initial begin
        $dumpfile(`DUMP_FILE);
        $dumpvars();

        a = 3;
        b = 1;
        #1
        `DO_TEST(zero, 'b0, "3 does not equals 1");
        `DO_TEST(negitive, 'b0, "3 is not less then 1");

        a = 2;
        b = 2;
        #1
        `DO_TEST(zero, 'b1, "2 equals 2");
        `DO_TEST(negitive, 'b0, "2 is not less then 2");

        a = 4;
        b = 5;
        #1
        `DO_TEST(zero, 'b0, "4 does not equals 5");
        `DO_TEST(negitive, 'b1, "4 is less then 5");

        $finish();
    end

endmodule
