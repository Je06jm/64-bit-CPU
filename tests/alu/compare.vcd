$date
	Thu Apr 22 19:35:56 2021
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module Arithmetic $end
$var wire 64 ! a [63:0] $end
$var wire 64 " b [63:0] $end
$var wire 1 # carryIn $end
$var wire 16 $ op [15:0] $end
$var reg 1 % carry $end
$var reg 1 & divByZero $end
$var reg 1 ' negitive $end
$var reg 64 ( result [63:0] $end
$var reg 64 ) sa [63:0] $end
$var reg 64 * sb [63:0] $end
$var reg 1 + zero $end
$upscope $end
$scope module ICompareTest $end
$var wire 1 , zero $end
$var wire 1 - negitive $end
$var reg 64 . a [63:0] $end
$var reg 64 / b [63:0] $end
$scope module cmp $end
$var wire 64 0 a [63:0] $end
$var wire 64 1 b [63:0] $end
$var reg 1 - negitive $end
$var reg 64 2 result [63:0] $end
$var reg 1 , zero $end
$upscope $end
$upscope $end
$scope module IValue $end
$var wire 64 3 in [63:0] $end
$var wire 1 4 signExtend $end
$var wire 4 5 size [3:0] $end
$var reg 64 6 out [63:0] $end
$upscope $end
$enddefinitions $end
#0
$dumpvars
bx 6
bz 5
z4
bz 3
b10 2
b1 1
b11 0
b1 /
b11 .
0-
0,
x+
bz *
bz )
bx (
x'
x&
x%
bz $
z#
bz "
bz !
$end
#1
1,
b0 2
b10 /
b10 1
b10 .
b10 0
#2
1-
0,
b1111111111111111111111111111111111111111111111111111111111111111 2
b101 /
b101 1
b100 .
b100 0
#3
