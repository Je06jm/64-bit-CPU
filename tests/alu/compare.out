#! /c/Source/iverilog-install/bin/vvp
:ivl_version "12.0 (devel)" "(s20150603-1110-g18392a46)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "C:\iverilog\lib\ivl\system.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_sys.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_textio.vpi";
:vpi_module "C:\iverilog\lib\ivl\v2005_math.vpi";
:vpi_module "C:\iverilog\lib\ivl\va_math.vpi";
:vpi_module "C:\iverilog\lib\ivl\v2009.vpi";
S_000001a8d01c74e0 .scope package, "$unit" "$unit" 2 1;
 .timescale 0 0;
S_000001a8d01c7670 .scope package, "instructions" "instructions" 3 7;
 .timescale 0 0;
enum000001a8d014ac70 .enum4 (4)
   "BITS_8" 4'b0000,
   "BITS_16" 4'b0001,
   "BITS_32" 4'b0010,
   "BITS_64" 4'b0011
 ;
enum000001a8d014bd30 .enum4 (16)
   "NOP" 16'b0000000000000000,
   "HALT" 16'b0000000000000001,
   "LOADMCR" 16'b0000000000000010,
   "STOREMCR" 16'b0000000000000011,
   "SYSCALL" 16'b0000000000000100,
   "SYSRET" 16'b0000000000000101,
   "INT" 16'b0000000000000110,
   "INTRET" 16'b0000000000000111,
   "ADD" 16'b0000000000001000,
   "SUB" 16'b0000000000001001,
   "MUL" 16'b0000000000001010,
   "UDIV" 16'b0000000000001011,
   "UMOD" 16'b0000000000001100,
   "SDIV" 16'b0000000000001101,
   "SMOD" 16'b0000000000001110,
   "INC" 16'b0000000000001111,
   "DEC" 16'b0000000000010000,
   "FADD" 16'b0000000000010001,
   "FSUB" 16'b0000000000010010,
   "FMUL" 16'b0000000000010011,
   "FDIV" 16'b0000000000010100,
   "FMOD" 16'b0000000000010101,
   "FFMA" 16'b0000000000010110,
   "FEPOW" 16'b0000000000010111,
   "F2POW" 16'b0000000000011000,
   "F10POW" 16'b0000000000011001,
   "FEPOW1" 16'b0000000000011010,
   "F2POW1" 16'b0000000000011011,
   "F10POW1" 16'b0000000000011100,
   "FLN" 16'b0000000000011101,
   "F2LOG" 16'b0000000000011110,
   "F10LOG" 16'b0000000000011111,
   "FLN1" 16'b0000000000100000,
   "F2LOG1" 16'b0000000000100001,
   "F10LOG1" 16'b0000000000100010,
   "FDISTANCE" 16'b0000000000100011,
   "FSQRT" 16'b0000000000100100,
   "FPOW1" 16'b0000000000100101,
   "FFPOW" 16'b0000000000100110,
   "FPOW" 16'b0000000000100111,
   "FSIN" 16'b0000000000101000,
   "FCOS" 16'b0000000000101001,
   "FTAN" 16'b0000000000101010,
   "FASIN" 16'b0000000000101011,
   "FACOS" 16'b0000000000101100,
   "FATAN" 16'b0000000000101101,
   "FATAN2" 16'b0000000000101110,
   "FSINPI" 16'b0000000000101111,
   "FCOSPI" 16'b0000000000110000,
   "FTANPI" 16'b0000000000110001,
   "FASINPI" 16'b0000000000110010,
   "FACOSPI" 16'b0000000000110011,
   "FATANPI" 16'b0000000000110100,
   "FATAN2PI" 16'b0000000000110101,
   "FSINH" 16'b0000000000110110,
   "FCOSH" 16'b0000000000110111,
   "FTANH" 16'b0000000000111000,
   "FARSINH" 16'b0000000000111001,
   "FARCOSH" 16'b0000000000111010,
   "FARTANH" 16'b0000000000111011,
   "FABS" 16'b0000000000111100,
   "FNEG" 16'b0000000000111101,
   "FCOPY" 16'b0000000000111110,
   "FTEST" 16'b0000000000111111,
   "TOINT" 16'b0000000001000000,
   "TOFLOAT" 16'b0000000001000001,
   "AND" 16'b0000000001000010,
   "OR" 16'b0000000001000011,
   "XOR" 16'b0000000001000100,
   "NOT" 16'b0000000001000101,
   "ROLR" 16'b0000000001000110,
   "ROLL" 16'b0000000001000111,
   "SHIFTR" 16'b0000000001001000,
   "SHIFTL" 16'b0000000001001001,
   "FLIP" 16'b0000000001001010,
   "SWAP" 16'b0000000001001011,
   "MOVE" 16'b0000000001001100,
   "CMP" 16'b0000000001001101,
   "JMP" 16'b0000000001001110,
   "BRA" 16'b0000000001001111,
   "RET" 16'b0000000001010000,
   "LOAD" 16'b0000000001010001,
   "STORE" 16'b0000000001010010,
   "STOREIF" 16'b0000000001010011,
   "PUSH" 16'b0000000001010100,
   "POP" 16'b0000000001010101,
   "BIT" 16'b0000000001010110,
   "OUT" 16'b0000000001010111,
   "IN" 16'b0000000001011000,
   "SETF" 16'b0000000001011001,
   "CLEARF" 16'b0000000001011010
 ;
S_000001a8d01acae0 .scope package, "types" "types" 4 4;
 .timescale 0 0;
enum000001a8d01b0ef0 .enum4 (2)
   "HALF" 2'b00,
   "SINGLE" 2'b01,
   "DOUBLE" 2'b10
 ;
S_000001a8d01acc70 .scope module, "Arithmetic" "Arithmetic" 5 7;
 .timescale 0 0;
    .port_info 0 /INPUT 16 "op";
    .port_info 1 /INPUT 1 "carryIn";
    .port_info 2 /INPUT 64 "a";
    .port_info 3 /INPUT 64 "b";
    .port_info 4 /OUTPUT 64 "result";
    .port_info 5 /OUTPUT 1 "divByZero";
    .port_info 6 /OUTPUT 1 "zero";
    .port_info 7 /OUTPUT 1 "carry";
    .port_info 8 /OUTPUT 1 "negitive";
o000001a8d01cf268 .functor BUFZ 64, C4<zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz>; HiZ drive
v000001a8d014d790_0 .net "a", 63 0, o000001a8d01cf268;  0 drivers
o000001a8d01cf298 .functor BUFZ 64, C4<zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz>; HiZ drive
v000001a8d014d830_0 .net "b", 63 0, o000001a8d01cf298;  0 drivers
v000001a8d014d8d0_0 .var "carry", 0 0;
o000001a8d01cf2f8 .functor BUFZ 1, C4<z>; HiZ drive
v000001a8d014d970_0 .net "carryIn", 0 0, o000001a8d01cf2f8;  0 drivers
v000001a8d014da10_0 .var "divByZero", 0 0;
v000001a8d019d640_0 .var "negitive", 0 0;
o000001a8d01cf388 .functor BUFZ 16, C4<zzzzzzzzzzzzzzzz>; HiZ drive
v000001a8d019d6e0_0 .net "op", 15 0, o000001a8d01cf388;  0 drivers
v000001a8d019d780_0 .var "result", 63 0;
v000001a8d019d820_0 .var/s "sa", 63 0;
v000001a8d019d8c0_0 .var/s "sb", 63 0;
v000001a8d019d960_0 .var "zero", 0 0;
E_000001a8d01bf720/0 .event anyedge, v000001a8d014d790_0, v000001a8d014d830_0, v000001a8d019d780_0, v000001a8d019d6e0_0;
E_000001a8d01bf720/1 .event anyedge, v000001a8d014d970_0, v000001a8d019d820_0, v000001a8d019d8c0_0;
E_000001a8d01bf720 .event/or E_000001a8d01bf720/0, E_000001a8d01bf720/1;
S_000001a8d01ace00 .scope module, "ICompareTest" "ICompareTest" 6 7;
 .timescale 0 0;
v000001a8d022a960_0 .var "a", 63 0;
v000001a8d022a8c0_0 .var "b", 63 0;
v000001a8d022a6e0_0 .net "negitive", 0 0, v000001a8d022a780_0;  1 drivers
v000001a8d022a320_0 .net "zero", 0 0, v000001a8d022ab40_0;  1 drivers
S_000001a8d01a3f00 .scope module, "cmp" "ICompare" 6 12, 7 5 0, S_000001a8d01ace00;
 .timescale 0 0;
    .port_info 0 /INPUT 64 "a";
    .port_info 1 /INPUT 64 "b";
    .port_info 2 /OUTPUT 1 "zero";
    .port_info 3 /OUTPUT 1 "negitive";
v000001a8d022abe0_0 .net "a", 63 0, v000001a8d022a960_0;  1 drivers
v000001a8d022adc0_0 .net "b", 63 0, v000001a8d022a8c0_0;  1 drivers
v000001a8d022a780_0 .var "negitive", 0 0;
v000001a8d022a0a0_0 .var "result", 63 0;
v000001a8d022ab40_0 .var "zero", 0 0;
E_000001a8d01bf760 .event anyedge, v000001a8d022abe0_0, v000001a8d022adc0_0, v000001a8d022a0a0_0;
S_000001a8d014d600 .scope module, "IValue" "IValue" 8 7;
 .timescale 0 0;
    .port_info 0 /INPUT 64 "in";
    .port_info 1 /OUTPUT 64 "out";
    .port_info 2 /INPUT 4 "size";
    .port_info 3 /INPUT 1 "signExtend";
o000001a8d01cf7d8 .functor BUFZ 64, C4<zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz>; HiZ drive
v000001a8d022afa0_0 .net "in", 63 0, o000001a8d01cf7d8;  0 drivers
v000001a8d022ac80_0 .var "out", 63 0;
o000001a8d01cf838 .functor BUFZ 1, C4<z>; HiZ drive
v000001a8d022a640_0 .net "signExtend", 0 0, o000001a8d01cf838;  0 drivers
o000001a8d01cf868 .functor BUFZ 4, C4<zzzz>; HiZ drive
v000001a8d022a820_0 .net "size", 3 0, o000001a8d01cf868;  0 drivers
E_000001a8d01beaa0 .event anyedge, v000001a8d022a820_0, v000001a8d022a640_0, v000001a8d022afa0_0;
    .scope S_000001a8d01acc70;
T_0 ;
    %wait E_000001a8d01bf720;
    %load/vec4 v000001a8d014d790_0;
    %assign/vec4 v000001a8d019d820_0, 0;
    %load/vec4 v000001a8d014d830_0;
    %assign/vec4 v000001a8d019d8c0_0, 0;
    %load/vec4 v000001a8d019d780_0;
    %pushi/vec4 0, 0, 64;
    %cmp/e;
    %flag_get/vec4 4;
    %assign/vec4 v000001a8d019d960_0, 0;
    %load/vec4 v000001a8d019d780_0;
    %parti/s 1, 63, 7;
    %assign/vec4 v000001a8d019d640_0, 0;
    %load/vec4 v000001a8d019d6e0_0;
    %pushi/vec4 13, 0, 16;
    %cmp/e;
    %flag_get/vec4 4;
    %load/vec4 v000001a8d019d6e0_0;
    %pushi/vec4 11, 0, 16;
    %cmp/e;
    %flag_get/vec4 4;
    %or;
    %load/vec4 v000001a8d019d6e0_0;
    %pushi/vec4 14, 0, 16;
    %cmp/e;
    %flag_get/vec4 4;
    %or;
    %load/vec4 v000001a8d019d6e0_0;
    %pushi/vec4 12, 0, 16;
    %cmp/e;
    %flag_get/vec4 4;
    %or;
    %load/vec4 v000001a8d014d830_0;
    %pushi/vec4 0, 0, 64;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
    %assign/vec4 v000001a8d014da10_0, 0;
    %load/vec4 v000001a8d019d6e0_0;
    %dup/vec4;
    %pushi/vec4 8, 0, 16;
    %cmp/u;
    %jmp/1 T_0.0, 6;
    %dup/vec4;
    %pushi/vec4 9, 0, 16;
    %cmp/u;
    %jmp/1 T_0.1, 6;
    %dup/vec4;
    %pushi/vec4 10, 0, 16;
    %cmp/u;
    %jmp/1 T_0.2, 6;
    %dup/vec4;
    %pushi/vec4 11, 0, 16;
    %cmp/u;
    %jmp/1 T_0.3, 6;
    %dup/vec4;
    %pushi/vec4 12, 0, 16;
    %cmp/u;
    %jmp/1 T_0.4, 6;
    %dup/vec4;
    %pushi/vec4 13, 0, 16;
    %cmp/u;
    %jmp/1 T_0.5, 6;
    %dup/vec4;
    %pushi/vec4 14, 0, 16;
    %cmp/u;
    %jmp/1 T_0.6, 6;
    %dup/vec4;
    %pushi/vec4 15, 0, 16;
    %cmp/u;
    %jmp/1 T_0.7, 6;
    %dup/vec4;
    %pushi/vec4 16, 0, 16;
    %cmp/u;
    %jmp/1 T_0.8, 6;
    %jmp T_0.9;
T_0.0 ;
    %load/vec4 v000001a8d014d790_0;
    %pad/u 65;
    %load/vec4 v000001a8d014d830_0;
    %pad/u 65;
    %add;
    %load/vec4 v000001a8d014d970_0;
    %pad/u 65;
    %add;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.1 ;
    %load/vec4 v000001a8d014d790_0;
    %pad/u 65;
    %load/vec4 v000001a8d014d830_0;
    %pad/u 65;
    %sub;
    %load/vec4 v000001a8d014d970_0;
    %pad/u 65;
    %sub;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.2 ;
    %load/vec4 v000001a8d014d790_0;
    %pad/u 65;
    %load/vec4 v000001a8d014d830_0;
    %pad/u 65;
    %mul;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.3 ;
    %load/vec4 v000001a8d014d790_0;
    %pad/u 65;
    %load/vec4 v000001a8d014d830_0;
    %pad/u 65;
    %div;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.4 ;
    %load/vec4 v000001a8d014d790_0;
    %pad/u 65;
    %load/vec4 v000001a8d014d830_0;
    %pad/u 65;
    %mod;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.5 ;
    %load/vec4 v000001a8d019d820_0;
    %pad/s 65;
    %load/vec4 v000001a8d019d8c0_0;
    %pad/s 65;
    %div/s;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.6 ;
    %load/vec4 v000001a8d019d820_0;
    %pad/s 65;
    %load/vec4 v000001a8d019d8c0_0;
    %pad/s 65;
    %mod/s;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.7 ;
    %load/vec4 v000001a8d014d790_0;
    %pad/u 65;
    %addi 1, 0, 65;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.8 ;
    %load/vec4 v000001a8d014d790_0;
    %pad/u 65;
    %subi 1, 0, 65;
    %split/vec4 64;
    %assign/vec4 v000001a8d019d780_0, 0;
    %assign/vec4 v000001a8d014d8d0_0, 0;
    %jmp T_0.9;
T_0.9 ;
    %pop/vec4 1;
    %jmp T_0;
    .thread T_0, $push;
    .scope S_000001a8d01a3f00;
T_1 ;
    %wait E_000001a8d01bf760;
    %load/vec4 v000001a8d022abe0_0;
    %load/vec4 v000001a8d022adc0_0;
    %sub;
    %assign/vec4 v000001a8d022a0a0_0, 0;
    %load/vec4 v000001a8d022a0a0_0;
    %pushi/vec4 0, 0, 64;
    %cmp/e;
    %flag_get/vec4 4;
    %assign/vec4 v000001a8d022ab40_0, 0;
    %load/vec4 v000001a8d022a0a0_0;
    %parti/s 1, 63, 7;
    %assign/vec4 v000001a8d022a780_0, 0;
    %jmp T_1;
    .thread T_1, $push;
    .scope S_000001a8d01ace00;
T_2 ;
    %pushi/vec4 0, 0, 64;
    %store/vec4 v000001a8d022a960_0, 0, 64;
    %pushi/vec4 0, 0, 64;
    %store/vec4 v000001a8d022a8c0_0, 0, 64;
    %end;
    .thread T_2;
    .scope S_000001a8d01ace00;
T_3 ;
    %vpi_call/w 6 23 "$dumpfile", "tests/alu/compare.vcd" {0 0 0};
    %vpi_call/w 6 24 "$dumpvars" {0 0 0};
    %pushi/vec4 3, 0, 64;
    %store/vec4 v000001a8d022a960_0, 0, 64;
    %pushi/vec4 1, 0, 64;
    %store/vec4 v000001a8d022a8c0_0, 0, 64;
    %delay 1, 0;
    %load/vec4 v000001a8d022a320_0;
    %pad/u 32;
    %cmpi/e 0, 0, 32;
    %jmp/0xz  T_3.0, 4;
    %vpi_call/w 6 31 "$display", "Passed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000011101, $realtime, "3 does not equals 1" {0 0 0};
    %jmp T_3.1;
T_3.0 ;
    %vpi_call/w 6 33 "$display", "Failed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000011101, $realtime, "3 does not equals 1" {0 0 0};
T_3.1 ;
    %load/vec4 v000001a8d022a6e0_0;
    %pad/u 32;
    %cmpi/e 0, 0, 32;
    %jmp/0xz  T_3.2, 4;
    %vpi_call/w 6 32 "$display", "Passed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000011110, $realtime, "3 is not less then 1" {0 0 0};
    %jmp T_3.3;
T_3.2 ;
    %vpi_call/w 6 34 "$display", "Failed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000011110, $realtime, "3 is not less then 1" {0 0 0};
T_3.3 ;
    %pushi/vec4 2, 0, 64;
    %store/vec4 v000001a8d022a960_0, 0, 64;
    %pushi/vec4 2, 0, 64;
    %store/vec4 v000001a8d022a8c0_0, 0, 64;
    %delay 1, 0;
    %load/vec4 v000001a8d022a320_0;
    %pad/u 32;
    %cmpi/e 1, 0, 32;
    %jmp/0xz  T_3.4, 4;
    %vpi_call/w 6 37 "$display", "Passed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000100011, $realtime, "2 equals 2" {0 0 0};
    %jmp T_3.5;
T_3.4 ;
    %vpi_call/w 6 39 "$display", "Failed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000100011, $realtime, "2 equals 2" {0 0 0};
T_3.5 ;
    %load/vec4 v000001a8d022a6e0_0;
    %pad/u 32;
    %cmpi/e 0, 0, 32;
    %jmp/0xz  T_3.6, 4;
    %vpi_call/w 6 38 "$display", "Passed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000100100, $realtime, "2 is not less then 2" {0 0 0};
    %jmp T_3.7;
T_3.6 ;
    %vpi_call/w 6 40 "$display", "Failed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000100100, $realtime, "2 is not less then 2" {0 0 0};
T_3.7 ;
    %pushi/vec4 4, 0, 64;
    %store/vec4 v000001a8d022a960_0, 0, 64;
    %pushi/vec4 5, 0, 64;
    %store/vec4 v000001a8d022a8c0_0, 0, 64;
    %delay 1, 0;
    %load/vec4 v000001a8d022a320_0;
    %pad/u 32;
    %cmpi/e 0, 0, 32;
    %jmp/0xz  T_3.8, 4;
    %vpi_call/w 6 43 "$display", "Passed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000101001, $realtime, "4 does not equals 5" {0 0 0};
    %jmp T_3.9;
T_3.8 ;
    %vpi_call/w 6 45 "$display", "Failed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000101001, $realtime, "4 does not equals 5" {0 0 0};
T_3.9 ;
    %load/vec4 v000001a8d022a6e0_0;
    %pad/u 32;
    %cmpi/e 1, 0, 32;
    %jmp/0xz  T_3.10, 4;
    %vpi_call/w 6 44 "$display", "Passed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000101010, $realtime, "4 is less then 5" {0 0 0};
    %jmp T_3.11;
T_3.10 ;
    %vpi_call/w 6 46 "$display", "Failed (%s, %0d [%0d]): %s", "tests/alu/compare.sv", 32'sb00000000000000000000000000101010, $realtime, "4 is less then 5" {0 0 0};
T_3.11 ;
    %vpi_call/w 6 44 "$finish" {0 0 0};
    %end;
    .thread T_3;
    .scope S_000001a8d014d600;
T_4 ;
    %wait E_000001a8d01beaa0;
    %load/vec4 v000001a8d022a820_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 4;
    %cmp/u;
    %jmp/1 T_4.0, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 4;
    %cmp/u;
    %jmp/1 T_4.1, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 4;
    %cmp/u;
    %jmp/1 T_4.2, 6;
    %dup/vec4;
    %pushi/vec4 3, 0, 4;
    %cmp/u;
    %jmp/1 T_4.3, 6;
    %jmp T_4.4;
T_4.0 ;
    %load/vec4 v000001a8d022a640_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_4.5, 8;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 1, 7, 4;
    %replicate 56;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 8, 0, 2;
    %concat/vec4; draw_concat_vec4
    %assign/vec4 v000001a8d022ac80_0, 0;
    %jmp T_4.6;
T_4.5 ;
    %pushi/vec4 0, 0, 56;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 8, 0, 2;
    %concat/vec4; draw_concat_vec4
    %assign/vec4 v000001a8d022ac80_0, 0;
T_4.6 ;
    %jmp T_4.4;
T_4.1 ;
    %load/vec4 v000001a8d022a640_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_4.7, 8;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 1, 15, 5;
    %replicate 48;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 16, 0, 2;
    %concat/vec4; draw_concat_vec4
    %assign/vec4 v000001a8d022ac80_0, 0;
    %jmp T_4.8;
T_4.7 ;
    %pushi/vec4 0, 0, 48;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 16, 0, 2;
    %concat/vec4; draw_concat_vec4
    %assign/vec4 v000001a8d022ac80_0, 0;
T_4.8 ;
    %jmp T_4.4;
T_4.2 ;
    %load/vec4 v000001a8d022a640_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_4.9, 8;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 1, 31, 6;
    %replicate 32;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 32, 0, 2;
    %concat/vec4; draw_concat_vec4
    %assign/vec4 v000001a8d022ac80_0, 0;
    %jmp T_4.10;
T_4.9 ;
    %pushi/vec4 0, 0, 32;
    %load/vec4 v000001a8d022afa0_0;
    %parti/s 32, 0, 2;
    %concat/vec4; draw_concat_vec4
    %assign/vec4 v000001a8d022ac80_0, 0;
T_4.10 ;
    %jmp T_4.4;
T_4.3 ;
    %load/vec4 v000001a8d022afa0_0;
    %assign/vec4 v000001a8d022ac80_0, 0;
    %jmp T_4.4;
T_4.4 ;
    %pop/vec4 1;
    %jmp T_4;
    .thread T_4, $push;
# The file index is used to find the file name in the following table.
:file_names 9;
    "N/A";
    "<interactive>";
    "-";
    "./rtl/instructions.svh";
    "./rtl/types.svh";
    "rtl/alu/arithmetic.sv";
    "tests/alu/compare.sv";
    "rtl/alu/compare.sv";
    "rtl/alu/value.sv";
